drivers: udc: disable SOF interrupt by default

jfischer-no opened the pull request:
Add UDC Kconfig option to enable SOF interrupts.
Disable SOF interrupt in drivers that are maintainable and do not obfuscate how interrupts are enabled. If the new Kconfig option is disabled, no SOF events are passed to the higher layer.

Resolves: #87732

I also wanted to address #74058 in the same PR, but that requires more upfront work.

jfischer-no committed:
drivers: udc: disable SOF interrupt by default

If the new Kconfig option is disabled, no SOF events are passed to the
higher layer.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

henrikbrixandersen commented:
> Disable SOF interrupt in drivers that are maintainable and do not obfuscate how interrupts are enabled.

Could you elaborate on this?

MarkWangChinese commented:
Hi @jfischer-no Could I add one commit in your pr to disable USB_DEVICE_CONFIG_SOF_NOTIFICATIONS based on CONFIG_UDC_ENABLE_SOF for NXP EHCI and IP3511 controller wrapper driver.

tmon-nordic commented on samples/subsys/usb/uac2_explicit_feedback/prj.conf:
```diff
@@ -3,6 +3,7 @@ CONFIG_I2S=y
 #USB related configs
 CONFIG_USB_DEVICE_STACK_NEXT=y
 CONFIG_USBD_AUDIO2_CLASS=y
+CONFIG_UDC_ENABLE_SOF=y
```
Should the class depend on `CONFIG_UDC_ENABLE_SOF`? UAC2 cannot work at all without SOF. Changing the default here just makes the actual samples work, but does not prevent in any way configuring an completely not working software.

jfischer-no replied:
I can add `depends on UDC_ENABLE_SOF`. I was not sure when I finished the commit. I am thinking about moving the SoF call to be in driver thread context, which could also be meta IRQ or have its own SoF meta IRQ thread, but that needs to be per usbd_context or have its own message queue and is IMO too much. Unfortunately, few drivers use the UDC work queue.

tmon-nordic replied:
I don't quite think meta IRQ is the way to go. One very important feature of SOF is that if the SW is too late, then all the past work does not really matter. The SW has always to worry only what to do within current SOF and the fact how many was missed is not really important.

Using queue for deferring SOF handling allows the queue to become full. I don't think it is about ensuring how fast the queue is handled, but rather about how to avoid the now-obsolete (past deadline) work.

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -2154,8 +2154,8 @@ static int udc_dwc2_init_controller(const struct device *dev)
 		    USB_DWC2_GINTSTS_WKUPINT | USB_DWC2_GINTSTS_USBSUSP |
-		    USB_DWC2_GINTSTS_INCOMPISOOUT | USB_DWC2_GINTSTS_INCOMPISOIN |
-		    USB_DWC2_GINTSTS_SOF,
+		    IF_ENABLED(CONFIG_UDC_ENABLE_SOF, (USB_DWC2_GINTSTS_SOF |))
+		    USB_DWC2_GINTSTS_INCOMPISOOUT | USB_DWC2_GINTSTS_INCOMPISOIN,
```
Isochronous transfers won't work without SOF and therefore there is no point in having INCOMPISOOUT and INCOMPISOIN enabled without SOF.

jfischer-no replied:
I don't quite understand. Will these interrupts be triggered even if SOF is disabled and must not be enabled here?

tmon-nordic replied:
INCOMPISOIN/INCOMPISOOUT will be triggered if SOF is disabled if an isochronous transfer was scheduled and host did not send the IN or OUT token within first 80% of the (micro-)frame.

Isochronous transfers won't really work on DWC2 without SOF interrupt due to how the scheudling works and that this manual reschedule is required from driver.

tmon-nordic commented on subsys/usb/device_next/class/Kconfig.uac2:
```diff
@@ -4,6 +4,7 @@
 
 config USBD_AUDIO2_CLASS
 	bool "USB Audio 2 class support [EXPERIMENTAL]"
+	depends on UDC_ENABLE_SOF
```
Actually `select UDC_ENABLE_SOF` is better than `depends on UDC_ENABLE_SOF` because:

- USB Audio 2 class support is then always visible in "New USB device stack" menu
- The requirement is reflected by showing `-*- SOF interrupt processing` (can only be disabled if UAC2 is disabled)

tmon-nordic replied:
This needs to be addressed.

jfischer-no replied:
I changed it to 'select' for now, but using 'depends on' is just fine and follows https://docs.zephyrproject.org/latest/build/kconfig/tips.html.  We've already experienced a few terrible 'select -> depends on' changes in the past.

jfischer-no committed:
drivers: udc: add SOF Kconfig option and SOF event helper

Add helper to handle SOF interrupts/events and new Kconfig option to
disable SOF interrupt.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on drivers/usb/udc/udc_common.h:
```diff
@@ -156,6 +156,50 @@ int udc_submit_event(const struct device *dev,
+ * @brief Helper function to store UDC SOF stamp.
+ *
+ * @param[in] dev     Pointer to device struct of the driver instance
+ * @param[in] fnumber Frame number
```
While for Full-Speed this is enough, the description is insufficient regarding to what value is expected at High-Speed.

Universal Serial Bus Specification Revision 2.0 10.2.3 Frame and Microframe Generation
> Host controllers operating with full-speed devices maintain a current frame number (at least 11 bits) that
increments at a 1 ms period. The host transmits the lower 11 bits of the current frame number in each SOF
token transmission.
Host controllers operating with high-speed devices maintain a current microframe number (at least 14 bits)
that increments at a 125 µs period. The host transmits bits 3 through 13 of the current microframe number
in each SOF token transmission. This results in the same SOF packet value being transmitted for eight
consecutive microframes before the SOF packet value increments.

Is the fnumber parameter supposed to be the actual microframe number (i.e. contain the bits 0 through 2, derived by device controller) when operating at High-Speed?

tmon-nordic commented on include/zephyr/drivers/usb/udc.h:
```diff
@@ -291,6 +291,10 @@ struct udc_data {
 	struct net_buf *setup;
 	/** Driver private data */
 	void *priv;
+#if CONFIG_UDC_ENABLE_SOF
```
Can this new functionality be split away from this PR?

tmon-nordic commented on subsys/usb/device_next/usbd_core.c:
```diff
@@ -202,6 +196,24 @@ static void usbd_thread(void *p1, void *p2, void *p3)
+{
+	struct usbd_context *const uds_ctx = (void *)udc_get_event_ctx(event->dev);
+
+	/*
+	 * Queue the SOF events only if they are from the ISR context.
+	 * Primarily to avoid message queue flooding from high speed
+	 * controllers.
+	 */
+	if (event->type == UDC_EVT_SOF && !k_is_in_isr()) {
+		usbd_class_bcast_event(uds_ctx, event);
+		return 0;
+	}
```
What's the real point here? Isn't this just guarding against broken drivers? Shouldn't this be an assert?

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -3177,6 +3178,11 @@ static ALWAYS_INLINE void dwc2_thread_handler(void *const arg)
 	}
 
+	if (evt & BIT(DWC2_DRV_EVT_SOF)) {
+		udc_update_sof_stamp(dev, priv->sof_num);
```
This doesn't make sense. The timestamp will have a way too big jitter for any practical use.
